pyvista/core/datasetattributes.py

```diff
@@ -1070,8 +1084,8 @@ def update(self, array_dict: Union[Dict[str, np.ndarray], 'DataSetAttributes']):
             self[name] = array.copy()
 
     def _raise_index_out_of_bounds(self, index: Any):
-        max_index = self.VTKObject.GetNumberOfArrays()
         if isinstance(index, int):
+            max_index = self.VTKObject.GetNumberOfArrays()
             if index < 0 or index >= self.VTKObject.GetNumberOfArrays():
                 raise KeyError(f'Array index ({index}) out of range [0, {max_index}]')
```

Comment:
When phrased like this, we should probably make this inclusive:
```suggestion
                raise KeyError(f'Array index ({index}) out of range [0, {max_index - 1}]')
```